contents/julia_101.md

```diff
@@ -1,40 +1,438 @@
+Supposed `jl name`'s birthday has passed and now it has turned `jl age+1`:
+
+```jl
+scob(
+"""
+age = 10
+"""
+)
+```
+
+We can do the same with its `name`, suppose that Julia has earned some titles due to its blazing speed.
+We would change the variable `name` to the new value:
+
+```jl
+scob(
+"""
+name = "Julia Scientificus Rapidus"
+"""
+)
+```
+
+We can also do operations on variables such as addition or division.
+Let's see how much months of age `jl name` has by multiplying `age` by 12:
+
+```jl
+scob(
+"""
+age * 12
+"""
+)
+```
+
+We can inspect variables types by using `typeof()` function:
+
+```{=comment}
+Rik: I have no idea why `scob` doesn't work here.
+It is overloaded somewhere in the wrong way it seems.
+```
+
+```jl
+sco("typeof(age)")
+```
+
+The next question the becomes: "What else can I do with `jl typeof(age)`?"
+There is a nice handy function `methodswith()` that spits out every function available, along with its signature, for a certain type.
+Here I will restrict the output only to the first 5 rows:
+
+```{=comment}
+Since the type is a Vector, Books converts it to multiple output statements.
+In this case, I've enforced to show the output of `Base.show`.
+```
+
+```jl
+sco(
+"""
+first(methodswith(Int64), 5)
+"""; post=Books.catch_show
+)
+```
+
+### User-defined Types {#sec:struct}
+
+Having variables around without any sort of hierarchy and relationships are not ideal.
+In Julia we can define that kind of structured data with a `struct` (also known as composite types).
+Inside each `struct` there is an optional set of fields.
+They differ from the primitive types (e.g. integer and floats) that are by default defined already inside the core of Julia language.
+Since most `struct`s are user-defined they are known as user-defined types.
+
+For example lets create a `struct` to represent scientific opensource programming languages.
+We'll also define a set of fields along with the corresponding types inside the `struct`:
+
+```jl
+sco("""
+struct Language
+    name::String
+    title::String
+    year_of_birth::Int64
+    fast::Bool
+end
+"""; post=x -> "")
+```
+
+To inspect the field names you can use the `fieldnames()` and passing the desired `struct` as an argument:
+
+```jl
+sco("fieldnames(Language)")
+```
+
+To use `struct`s, we must instantiate individual instances (or "objects"), each with its own specific values for the fields defined inside the `struct`.
+Let's instantiate two instances, one for Julia and one for Python with the appropriate types as fields inside the `Language()` constructor:
+
+```jl
+scob(
+"""
+julia = Language("Julia", "Scientificus Rapidus", 2012, true)
+python = Language("Python", "C/FORTRAN Dependentus Letargicus", 1991, false)
+"""
+)
+```
+
+One thing to note with `struct`s is that we cannot change their values once they are instantiated.
+We can solve this with a `mutable struct`.
+Also note that everything in Julia that we impose mutability will take a performance hit.
```

Comment:
```suggestion
Also, note that mutable objects will, generally, be slower and more error prone.
```
For clarity: I mean error prone because it's easy to update a mutable struct somewhere in one point in the codebase and not notice that it has been updated when looking using it at a completely different part of the codebase.

I say "generally" since there are probably situations where mutable objects are more quick, but I'm not sure

Reply:
Good Point!